sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.directconnectivity.rntbd;
+
+import com.azure.cosmos.implementation.GoneException;
+import com.azure.cosmos.implementation.RxDocumentServiceRequest;
+import com.azure.cosmos.implementation.directconnectivity.IAddressResolver;
+import com.azure.cosmos.implementation.routing.PartitionKeyRangeIdentity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.channels.ClosedChannelException;
+import java.time.Instant;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class RntbdConnectionStateListener {
+    // region Fields
+
+    private static final Logger logger = LoggerFactory.getLogger(RntbdConnectionStateListener.class);
+
+    private final IAddressResolver addressResolver;
+    private final RntbdEndpoint endpoint;
+    private final Set<PartitionKeyRangeIdentity> partitionAddressCache;
```

Comment:
`RntbdConnectionStateListener:partitionAddressCache`  is a local cache and can be outdated. as a result of that `RntbdConnectionStateListener` may remove addresses of unrelated pkranges causing unnecessary refresh.

Consider this scenario:
1. PKR1, PKR2, PKR3 are hosted on the same physical node.
2. request1 with PKR1, request2 with PKR2, and request3 with PKR3 are sent and as a result RntbdConnectionStateListener: partitionAddressCache is populated.
3. due to  partition movement PKR1 moves out of the physical node, but PKR2 and PKR3 stay on the node.
4. now the node hosting PKR2 and PKR3 shuts down.
5. RntbdConnectionStateListener thinks PKR1 is still on this physical node, and hence will correctly remove addresses of PKR1, PKR2, and PKR3. But PKR3 addresses shouldn't have been removed!!.

Reply:
Yea, based on the current implementation, it could happen. 
Will think more and check how .net implement and then update here.

Reply:
let's track this as an optimization for later. (out of the scope of this PR)

Reply:
Created github working item to track this issue:
1. https://github.com/Azure/azure-sdk-for-java/issues/16245
2. https://github.com/Azure/azure-sdk-for-java/issues/16246